### ISSUE TITLE
Lift tries to delete data from a relation on unrelated schema changes

### ISSUE THREAD

**Issue by deibeljc** [prisma/migrate#214]
I have an issue where after I have introspected my DB, lift wants to delete my relations regardless of what change I make to the schema.

When I make a change (for example aPIKey to APIKey), I see this warning.
```
There will be data loss:

     • You are about to alter the column `ecommerceCredentialsId` on the `Merchants` table,
which still contains 2 non-null values. The data in that column will be lost.
     • You are about to alter the column `integratedCredentialsId` on the `Merchants` table,
 which still contains 313 non-null values. The data in that column will be lost.
```

## Gist of the requested info.
https://gist.github.com/deibeljc/db7cb91b63fdc27c4e435fa561b0dae2

**Comment by tomhoule** [comment]
Just for context, the original slack conversation: https://prisma.slack.com/archives/CKQTGR6T0/p1573655535208600

**Comment by deibeljc** [comment]
I tested and this issue does not happen with preview15, so it must be a new issue that popped up in 16

**Comment by pantharshit00** [comment]
> Turns out preview-015 did this, it just didn't notify me about it beforehand!

Yes, destructive changes alerts were implemented in preview 16. I think this issue is related to the column being named as `MerhId` and there are also some issues around integer IDs right now.

I have able to create a minimal reproduction of this:

1. Use the following SQL to bootstrap a database:
```sql
-- -------------------------------------------------------------
-- TablePlus 2.11.2(278)
--
-- https://tableplus.com/
--
-- Database: issue214
-- Generation Time: 2019-11-18 16:49:25.8210
-- -------------------------------------------------------------


/*!40101 SET @OLD_CHARACTER_SET_CLIENT=@@CHARACTER_SET_CLIENT */;
/*!40101 SET @OLD_CHARACTER_SET_RESULTS=@@CHARACTER_SET_RESULTS */;
/*!40101 SET @OLD_COLLATION_CONNECTION=@@COLLATION_CONNECTION */;
/*!40101 SET NAMES utf8 */;
/*!40103 SET @OLD_TIME_ZONE=@@TIME_ZONE */;
/*!40103 SET TIME_ZONE='+00:00' */;
/*!40014 SET @OLD_UNIQUE_CHECKS=@@UNIQUE_CHECKS, UNIQUE_CHECKS=0 */;
/*!40014 SET @OLD_FOREIGN_KEY_CHECKS=@@FOREIGN_KEY_CHECKS, FOREIGN_KEY_CHECKS=0 */;
/*!40101 SET @OLD_SQL_MODE=@@SQL_MODE, SQL_MODE='NO_AUTO_VALUE_ON_ZERO' */;
/*!40111 SET @OLD_SQL_NOTES=@@SQL_NOTES, SQL_NOTES=0 */;


DROP TABLE IF EXISTS `table1`;
CREATE TABLE `table1` (
  `MerchID` bigint(20) unsigned NOT NULL AUTO_INCREMENT,
  `name` varchar(40) DEFAULT NULL,
  `table2Id` bigint(20) unsigned DEFAULT NULL,
  PRIMARY KEY (`MerchID`),
  UNIQUE KEY `id` (`MerchID`),
  KEY `table2Id` (`table2Id`),
  CONSTRAINT `table1_ibfk_1` FOREIGN KEY (`table2Id`) REFERENCES `table2` (`id`)
) ENGINE=InnoDB AUTO_INCREMENT=2 DEFAULT CHARSET=utf8;

DROP TABLE IF EXISTS `table2`;
CREATE TABLE `table2` (
  `id` bigint(20) unsigned NOT NULL AUTO_INCREMENT,
  `something` text,
  PRIMARY KEY (`id`),
  UNIQUE KEY `id` (`id`)
) ENGINE=InnoDB AUTO_INCREMENT=2 DEFAULT CHARSET=utf8;

INSERT INTO `table1` (`MerchID`, `name`, `table2Id`) VALUES ('1', 'name1', '1');

INSERT INTO `table2` (`id`, `something`) VALUES ('1', 'name2');




/*!40103 SET TIME_ZONE=@OLD_TIME_ZONE */;
/*!40101 SET SQL_MODE=@OLD_SQL_MODE */;
/*!40014 SET FOREIGN_KEY_CHECKS=@OLD_FOREIGN_KEY_CHECKS */;
/*!40014 SET UNIQUE_CHECKS=@OLD_UNIQUE_CHECKS */;
/*!40101 SET CHARACTER_SET_CLIENT=@OLD_CHARACTER_SET_CLIENT */;
/*!40101 SET CHARACTER_SET_RESULTS=@OLD_CHARACTER_SET_RESULTS */;
/*!40101 SET COLLATION_CONNECTION=@OLD_COLLATION_CONNECTION */;
/*!40111 SET SQL_NOTES=@OLD_SQL_NOTES */;
```

2. Introspect the database. 
3. Due to #https://github.com/prisma/prisma2/issues/778, the database will not print in the url so add that in the schema file
4. Run `prisma2 lift save` and `prisma2 lift up`


cc @tomhoule

**Comment by tomhoule** [comment]
This appears to be related to the more general problem of migration-after-introspection. Introspection will give you types that are currently valid in a `schema.prisma` like `String` and `Int`, but the original database types do not match those Lift would use (`TEXT` instead of `VARCHAR` for `table2.name`, `INTEGER` instead of `BIGINT` for the IDs, in the example given by Harshit), so it alters the columns to conform to the types _it_ would apply. Currently all `ALTER COLUMN`s will lose all the data in the column, hence the warnings.

There are two ways out of this:

1. Wait until we can represent types like `BIGINT` and `VARCHAR` in the `schema.prisma`. We are currently working on this.
2. Make SQL schema diffing in lift more lenient (so it would consider `VARCHAR` and `TEXT` to both be `String` for example.

We are probably going to go for 1. Maybe @mavilein knows more?

**Comment by tomhoule** [comment]
I investigated this for a few hours and found two separate issues where Lift is "forced" to migrate after introspecting with the schema above:

1. The column defaults. What lift is doing currently isn't optimal. Progress on this issue is blocked on https://github.com/prisma/specs/issues/308.
2. The `@relation` getting changed when from ON DELETE RESTRICT (database default) to ON DELETE SET NULL (prisma default) without a chance to set this right in your `prisma.schema`. Here we should see if we could make it possible, and otherwise wait for non-destructive foreign key changes.

I will discuss that last issue with @mavilein tomorrow because he knows more than me.

**Comment by tomhoule** [comment]
Update: the issue was discussed with @mavilein - this is currently blocked on spec work.

**Comment by divyenduz** [comment]
Internal note: dropping for now

**Comment by deibeljc** [comment]
@tomhoule @divyenduz will this be fixed before GA/prod ready? Also is there a possible workaround? I want to use prisma2, but this is a blocking issue for my main service right now

**Comment by tomhoule** [comment]
This will absolutely be fixed for the Lift release :)

**Comment by tomhoule** [comment]
I am fairly confident this was fixed a while ago and cannot be reproduced in `prisma2@alpha` anymore, I will close the issue - please feel free to reopen if the problem persists :)

**Comment by tomhoule** [comment]
Status update: the specs work blocking this mentioned in https://github.com/prisma/lift/issues/214#issuecomment-565114692 has not made progress yet.

**Comment by tomhoule** [comment]
We removed the implicit defaults recently, and `ON DELETE`s currently cannot be expressed in the schema and they are not diffed by the migration engine, so I think this issue can be closed. Since a lot has changed in the last few months, if you still experience problems, please create a new issue :)
